x/ccv/consumer/keeper/genesis.go

```diff
@@ -97,63 +103,97 @@ func (k Keeper) InitGenesis(ctx sdk.Context, state *consumertypes.GenesisState)
+			mat := types.MaturingVSCPacket{
 				VscId:        vscId,
 				MaturityTime: timeNs,
 			}
 			maturingPackets = append(maturingPackets, mat)
 			return false
-		}
-		k.IteratePacketMaturityTime(ctx, cb)
+		})
 
-		gs.MaturingPackets = maturingPackets
-		return gs
-	}
-	clientID, ok := k.GetProviderClientID(ctx)
-	// if provider clientID and channelID don't exist on the consumer chain, then CCV protocol is disabled for this chain
-	// return a disabled genesis state
-	if !ok {
-		return consumertypes.DefaultGenesisState()
-	}
-	cs, ok := k.clientKeeper.GetClientState(ctx, clientID)
-	if !ok {
-		panic("provider client not set on already running consumer chain")
-	}
-	tmCs, ok := cs.(*ibctmtypes.ClientState)
-	if !ok {
-		panic("provider client consensus state is not tendermint client state")
-	}
-	consState, ok := k.clientKeeper.GetLatestClientConsensusState(ctx, clientID)
-	if !ok {
-		panic("provider consensus state not set on already running consumer chain")
-	}
-	tmConsState, ok := consState.(*ibctmtypes.ConsensusState)
-	if !ok {
-		panic("provider consensus state is not tendermint consensus state")
+		heightToVCIDs := []types.HeightToValsetUpdateID{}
+		k.IterateHeightToValsetUpdateID(ctx, func(height, vscID uint64) bool {
+			hv := types.HeightToValsetUpdateID{
+				Height:         height,
+				ValsetUpdateId: vscID,
+			}
+			heightToVCIDs = append(heightToVCIDs, hv)
+			return true
+		})
+
+		outstandingDowntimes := []types.OutstandingDowntime{}
+		k.IterateOutstandingDowntime(ctx, func(addr string) bool {
+			od := types.OutstandingDowntime{
+				ValidatorConsensusAddress: addr,
+			}
+			outstandingDowntimes = append(outstandingDowntimes, od)
+			return false
+		})
+
+		genesis = types.NewRestartGenesisState(
+			clientID,
+			channelID,
+			maturingPackets,
+			valset,
+			heightToVCIDs,
+			outstandingDowntimes,
+			params,
+		)
+	} else {
+		clientID, ok := k.GetProviderClientID(ctx)
+		// if provider clientID and channelID don't exist on the consumer chain, then CCV protocol is disabled for this chain
+		// return a disabled genesis state
+		if !ok {
+			return consumertypes.DefaultGenesisState()
+		}
+		cs, ok := k.clientKeeper.GetClientState(ctx, clientID)
+		if !ok {
+			panic("provider client not set on already running consumer chain")
+		}
+		tmCs, ok := cs.(*ibctmtypes.ClientState)
+		if !ok {
+			panic("provider client consensus state is not tendermint client state")
+		}
+		consState, ok := k.clientKeeper.GetLatestClientConsensusState(ctx, clientID)
+		if !ok {
+			panic("provider consensus state not set on already running consumer chain")
+		}
+		tmConsState, ok := consState.(*ibctmtypes.ConsensusState)
+		if !ok {
+			panic("provider consensus state is not tendermint consensus state")
+		}
+		// export client states and pending slashing requests into a new chain genesis
+		genesis = consumertypes.NewInitialGenesisState(tmCs, tmConsState, valset, k.GetPendingSlashRequests(ctx), params)
```

Comment:
Also, probably we need to export also `HeightValsetUpdateID` for the pending slash requests to be valid after restart.

Reply:
Pending slash requests should all have the same valset update id equals  to 0 right ?

Reply:
You are right. I forgot that the pending slash requests already have the mapping done, i.e., https://github.com/cosmos/interchain-security/blob/22bc31091af7f59bddb3ac63fcecebbcf3766cf6/x/ccv/consumer/keeper/validators.go#L93.